Add build() method to SortBuilder implementations

cbuescher opened the pull request:
For the current refactoring of SortBuilders related to #10217,
each SortBuilder should get a build() method that produces a
SortField according to the SortBuilder parameters on the shard.

This change also slightly refactors the current parse method in
SortParseElement to extract an internal parse method that returns
a list of sort fields only needs a QueryShardContext as input
instead of a full SearchContext. This allows using this internal
parse method for testing.

cbuescher committed:
Add build() method to SortBuilder implementations

For the current refactoring of SortBuilders related to #10217,
each SortBuilder should get a build() method that produces a
SortField according to the SortBuilder parameters on the shard.

This change also slightly refactors the current parse method in
SortParseElement to extract an internal parse method that returns
a list of sort fields only needs a QueryShardContext as input
instead of a full SearchContext. This allows using this internal
parse method for testing.

cbuescher commented:
@MaineC could you take a look at this? The generic test added to AbstractSortTestCase doesn't do many assertions on the two SortFields it compares, maybe you have an idea how to add to those tests. Also, those tests will be temporary and will be obsolete once we remove the old SortParseElement.

MaineC commented on core/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java:
```diff
@@ -220,6 +234,56 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+                throw new QueryShardException(context, "Sorting not supported for field[" + fieldName + "]");
+            }
+
+            // Enable when we also know how to detect fields that do tokenize, but only emit one token
```
Instead of having code commented out here - would it be better to track that in some follow up issue? (I know the comment was there before the refactoring, but code that's commented out always smells strange IMHO)

cbuescher replied:
Just talked to @martijnvg, this is old code and can be removed.

MaineC commented on core/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java:
```diff
@@ -220,6 +234,56 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+            if (sortMode != null) {
+                localSortMode = MultiValueMode.fromString(sortMode.toString());
+            }
+            if (fieldType.isNumeric() == false && (sortMode == SortMode.SUM || sortMode == SortMode.AVG)) {
```
Lost one comment copying this over - intentional?

MaineC commented on core/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java:
```diff
@@ -220,6 +234,56 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+            }
+            boolean reverse = (order == SortOrder.DESC);
+            if (localSortMode == null) {
+                localSortMode = reverse ? MultiValueMode.MAX : MultiValueMode.MIN;
```
By directly using order here we could delete the variable "reverse" above.

MaineC replied:
Hmpf - should first read through and comment afterwards. I saw only now that we need "reverse" further below anyway :/

MaineC commented on core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java:
```diff
@@ -461,7 +475,89 @@ public GeoDistanceSortBuilder fromXContent(QueryParseContext context, String ele
+
+        if (!indexCreatedBeforeV2_0 && !ignoreMalformed) {
+            for (GeoPoint point : localPoints) {
+                if (point.lat() > 90.0 || point.lat() < -90.0) {
```
Why not use this: [GeoUtils.isValidLatitude()](https://github.com/elastic/elasticsearch/blob/52ee4c70275a6b54ac193f5a13e185355dc962bc/core/src/main/java/org/elasticsearch/common/geo/GeoUtils.java#L79)

cbuescher replied:
sure

MaineC commented on core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java:
```diff
@@ -461,7 +475,89 @@ public GeoDistanceSortBuilder fromXContent(QueryParseContext context, String ele
+                if (point.lat() > 90.0 || point.lat() < -90.0) {
+                    throw new ElasticsearchParseException("illegal latitude value [{}] for [GeoDistanceSort]", point.lat());
+                }
+                if (point.lon() > 180.0 || point.lon() < -180) {
```
See above, there's another method for longitude validation.

cbuescher replied:
done

MaineC commented on core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java:
```diff
@@ -461,7 +475,89 @@ public GeoDistanceSortBuilder fromXContent(QueryParseContext context, String ele
+        }
+
+        if (sortMode == SortMode.SUM) {
+            throw new IllegalArgumentException("sort_mode [sum] isn't supported for sorting by geo distance");
```
https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-sort.html ... says that "geo distance sorting supports sort_mode options: min, max and avg.

Looking here: https://github.com/elastic/elasticsearch/blob/7ccf7adbdf748e3dc2cb37c027bcf30a1bae54a9/core/src/main/java/org/elasticsearch/search/sort/SortMode.java we have min, max, avg, sum and *median*

We forbid sum above - but what about median?

cbuescher replied:
Two things here, need to do some more digging: 
* MEDIAN seems to have been added just recently to MultiValueMode (a5c0ac0) 
* I see no tests for AVG or MEDIAN in GeoDistanceSortBuilderIT, so in order to see if those work we can probably add tests there

cbuescher replied:
MultiValueMode.MEDIAN seems to work with GeoDistanceSort, I added a test to GeoDistanceSortBuilderIT. The check for correct SortMode in build() can be removed because we already check this in the setter and there should be no other way to set the SortMode.

MaineC commented on core/src/main/java/org/elasticsearch/search/sort/ScriptSortBuilder.java:
```diff
@@ -244,6 +265,81 @@ public ScriptSortBuilder fromXContent(QueryParseContext context, String elementN
+            valueMode = reverse ? MultiValueMode.MAX : MultiValueMode.MIN;
+        }
+
+        final Nested nested = resolveNested(context, nestedPath, nestedFilter);
```
Huge +1 to factoring resolving nested out into it's own method.

MaineC commented on core/src/main/java/org/elasticsearch/search/sort/ScriptSortBuilder.java:
```diff
@@ -244,6 +265,81 @@ public ScriptSortBuilder fromXContent(QueryParseContext context, String elementN
+                };
+                break;
+            case NUMBER:
+                // TODO: should we rather sort missing values last?
```
I know this was a todo before, but would it make sense to track that in a separate issue to give it better visibility?

cbuescher replied:
I opened https://github.com/elastic/elasticsearch/issues/17190 and will remove this TODO here and from ScriptSortParser.

MaineC commented on core/src/main/java/org/elasticsearch/search/sort/ScriptSortParser.java:
```diff
@@ -140,7 +140,6 @@ public SortField parse(XContentParser parser, QueryShardContext context) throws
             sortMode = reverse ? MultiValueMode.MAX : MultiValueMode.MIN;
         }
 
-        // If nested_path is specified, then wrap the `fieldComparatorSource` in a `NestedFieldComparatorSource`
```
I don't quite get why this comment went away?

cbuescher replied:
The commit it was introduced (db421742f7) contained some code that I think it refers to, but that is no longer here. I might be mistaken, but I think its a left-over

cbuescher replied:
Oh, and NestedFieldComparatorSource doesn't seem to exist any more. Another sign that the comment might be outdated.

MaineC replied:
Makes sense to get rid of it then.

MaineC commented on core/src/test/java/org/elasticsearch/search/sort/FieldSortBuilderTests.java:
```diff
@@ -25,7 +25,7 @@
     @Override
     protected FieldSortBuilder createTestItem() {
-        String fieldName = randomAsciiOfLengthBetween(1, 10);
+        String fieldName = rarely() ? SortParseElement.DOC_FIELD_NAME : randomAsciiOfLengthBetween(1, 10);
```
good catch...

MaineC commented:
Overall looks awesome!

Left a few minor comments and questions.

About testing: I'm not sure we can test a whole lot more without knowing which type of sort we are dealing with. One idea (that would also be valid after removing the old parsing code) would be to add tests to the individual builders that have a few deeper assertions.

With this change it should already be possible to parse more than one sort element, no? This could be an addition to the tests.

cbuescher committed:
Adressing review comments

cbuescher commented:
@MaineC thanks, I adressed most of your comments, still need to look into the Median-SortMode support and maybe open an issue about sorting with missing values.

> About testing: I'm not sure we can test a whole lot more without knowing which type of sort we are dealing with. One idea (that would also be valid after removing the old parsing code) would be to add tests to the individual builders that have a few deeper assertions.

I will see if we can add a "assertSortField" method to each individual test that does some more inspection, maybe that works

> With this change it should already be possible to parse more than one sort element, no? This could be an addition to the tests.

No, the SortParseElement will parse a whole ` "sort" : {  }` block but the builders themselves currently only represent one sort option of which we can have many. This parsing / building needs to go somewhere up into the abstract SortBuilder or somewhere else I think. But I'd say that is a follow up PR.

cbuescher committed:
Adding tests for median sort mode to GeoDistanceSortBuilderIT

Also adding checks for median SortMode on non-numeric field types
to FieldSortBuilder, removing some unused code and switching
GeoDistanceSortBuilder to using ParseField.

cbuescher commented:
@MaineC I also adressed the rest of your comments, can you take another look if this looks good now?

MaineC commented:
Looked over your changes and the comments you added - LGTM

cbuescher committed:
Merge branch 'master' into sort-add-build

cbuescher committed:
Make sure to use nestedScope levels when building nested filters

cbuescher commented:
@MaineC thanks, I needed to move over few more lines from NestedInnerQueryParseSupport to the part in SortBuilder that is building the nested filters. I added those and merged in master once again.
